### ISSUE TITLE
Error has circular JSON structure

### ISSUE THREAD

**Issue by hrsvrma** [elastic/elasticsearch-js#848]
HI,

Error returned/thrown from @elastic/elasticsearch@6 has circular structure hence it is difficult to stringify using JSON.stringify or send in API response (i.e. using express.js). The Error in legacy version is not circular though.

TypeError: Converting circular structure to JSON
    at JSON.stringify (<anonymous>)

I am curious to know the reason behind circular structure of Error. And please suggest if there is a simple way to de-circle the Error circular structure.

Thanks

**Comment by delvedor** [comment]
Hello! You can't serialize circular structures, and you can't store a circular object in a database.
You should try to avoid circular objects directly in your code, and if you can't, you should override our default serializer with one that will handle circular objects, such as [fast-safe-stringify](https://www.npmjs.com/package/fast-safe-stringify).

```js
const stringify = require('fast-safe-stringify')
const { Client, Serializer } = require('@elastic/elasticsearch')

class CircularSerializer extends Serializer {
  serialize (obj) {
    return stringify(obj)
  }
}

const client = new Client({
  node: 'http://localhost:9200',
  Serializer: CircularSerializer
})
```

**Comment by delvedor** [comment]
Hello!
I'll open a pr to address this, but you should avoid sending the entire error message to your users, it will expose you to security vulnerabilities since the error object contains all the connection metadata and much more.

**Comment by delvedor** [comment]
Closed in #849.

**Comment by hrsvrma** [comment]
Hi Delvedor,

Thanks for the pr 😀.

I am not sending the entire error message to user as I too understand the security risk involved. This is just for development environment where I prefer to get error in API response instead of looking into server logs.
